src/vos/tests/vos_agg.c

```diff
@@ -0,0 +1,723 @@
+		}
+	}
+
+	return ops;
+}
+
+void
+pin_threads(void)
+{
+	ABT_thread	 child;
+	int		 i;
+
+	pin_thread((void *)(uintptr_t)0);
+	for (i = 1; i < DEFAULT_NUM_XSTREAMS; i++) {
+		ABT_thread_create(pools[i], pin_thread, (void *)(uintptr_t)i, ABT_THREAD_ATTR_NULL, &child);
+		ABT_thread_free(&child);
+	}
+}
+
+void run_bench(int num_init, int num_ops)
+{
+	int         i;
+	ABT_pool    target_pool = pools[1];
+	ABT_thread *children    = malloc(sizeof(*children) * (num_ops + 1));
+	struct agg_info	 agg_info = {0};
+	struct io_op	*args;
+	int		rc;
+	uint64_t	start, end;
+	double bw;
+
+	oid = dts_unit_oid_gen(0, DAOS_OF_DKEY_UINT64, 0);
+
+	if (gen_csum) {
+		rc = daos_csummer_init_with_type(&csummer, HASH_TYPE_CRC32, 1 << 12, 0);
+		D_ASSERT(rc == 0);
+	}
+
+	args = allocate_ops(num_ops, gen_csum);
+
+	for (i = 0; i < num_init; i++) {
+		ABT_thread_create(target_pool, submit_io, &args[i], ABT_THREAD_ATTR_NULL,
+				  &children[i]);
+	}
+
+	for (i = 0; i < num_init; i++) {
+		ABT_thread_free(&children[i]);
+	}
+
+	agg_info.epr.epr_lo = 0;
+	agg_info.epr.epr_hi = highest_io;
+
+	start = daos_get_ntime();
+	ABT_thread_create(target_pool, agg_thread, &agg_info, ABT_THREAD_ATTR_NULL,
+			  &children[num_ops]);
+
+	for (i = num_init; i < num_ops; i++) {
+		ABT_thread_create(target_pool, submit_io, &args[i], ABT_THREAD_ATTR_NULL,
+				  &children[i]);
+	}
+
+	end = start;
+	for (i = num_init; i < num_ops + 1; i++) {
+		if (i == num_ops)
+			end = daos_get_ntime();
+		ABT_thread_free(&children[i]);
+	}
+
+	bw = ((double)num_init * BUF_SIZE * NSEC_PER_SEC) / ((1024 * 1024) * agg_info.time_nsec);
+	printf("agg_time = %10.3lf ms, BW %10.5lf MB/s\n",
+	       (double)agg_info.time_nsec / NSEC_PER_MSEC, bw);
+	bw = ((double)(num_ops - num_init) * BUF_SIZE * NSEC_PER_SEC) /
+		(1024 * 1024 * (end - start));
+	printf("io_time  = %10.3lf ms, BW %10.5lf MB/s\n", (double)(end - start) / NSEC_PER_MSEC,
+	       bw);
+
+	if (gen_csum) {
+		daos_csummer_destroy(&csummer);
+	}
+}
+
+void
+print_usage(const char *name)
+{
+	printf("Usage: %s [opts]\n", name);
+	printf("\t-h            Print help and exit\n");
+	printf("\t-o count      Set number of operations to perform\n");
+	printf("\t-t d|D|n|N    Aggregation with(d) or without DSA(d), capitalize for csum, default is 'n'\n");
+	printf("\t-r ratio      Integer between 1 and 100 indicating percentage of total writes to aggregate\n");
+	printf("\t-p s|h	s is software DSA, h is hardware DSA\n");
+}
+
+/* Workaround for CART-890 */
+static void
+mem_pin_workaround(void)
+{
+        struct rlimit   rlim;
+        int             rc = 0;
+
+        /* Note: mallopt() returns 1 on success */
+        /* Prevent malloc from releasing memory via sbrk syscall */
+        rc = mallopt(M_TRIM_THRESHOLD, -1);
+        if (rc != 1)
+                D_WARN("Failed to disable malloc trim: %d\n", errno);
+
+        /* Disable fastbins; this option is not available on all systems */
+        rc = mallopt(M_MXFAST, 0); 
```

Comment:
(style)  trailing whitespace
(style)  code indent should use tabs where possible